src/mechanics/timeIntegration/NewmarkDirect.cpp

```diff
@@ -584,84 +248,320 @@ void NuTo::NewmarkDirect::PrintInfoStagger() const
+    ConstitutiveInputMap inputMap;
+    inputMap[Constitutive::eInput::CALCULATE_STATIC_DATA] =
+            std::make_unique<ConstitutiveCalculateStaticData>(eCalculateStaticData::EULER_BACKWARD);
+    inputMap.emplace(Constitutive::eInput::TIME, std::make_unique<ConstitutiveScalar>());
+
+    double& inputTime = (*inputMap.find(Constitutive::eInput::TIME)->second)[0];
+    inputTime = mTimeControl.GetCurrentTime();
+    return inputMap;
+}
+
+
+StructureOutputBlockVector NuTo::NewmarkDirect::EvaluateInternalGradient()
+{
+    const auto& dofStatus = mStructure->GetDofStatus();
+    StructureOutputMap evalInternalGradient;
+    StructureOutputBlockVector intForce(dofStatus, true);
+    evalInternalGradient[eStructureOutput::INTERNAL_GRADIENT] = &intForce;
+
+    auto inputMap = CreateInputMap();
+    mStructure->Evaluate(inputMap, evalInternalGradient);
+    return intForce;
+}
+
 
-    //    if (rHessian_dt0.IsConstant())
-    //    {
-    //        // save the values in rHessian0 for later evaluation and allocate a new hessian object
-    //        StructureOutputBlockMatrix hessian(mStructure->GetDofStatus());
-    //        hessian = rHessian_dt0;
-    //        if (mStructure->GetNumTimeDerivatives() >= 1)
-    //            hessian.AddScal(rHessian_dt1, mGamma / (mBeta * rTimeStep));
-    //
-    //        if (mStructure->GetNumTimeDerivatives() >= 2)
-    //            hessian.AddScal(rHessian_dt2, 1. / (mBeta * rTimeStep * rTimeStep));
-    //
-    //        hessian.ApplyCMatrix(mStructure->GetConstraintMatrix());
-    //        return mStructure->SolveBlockSystem(hessian.JJ, rResidualMod);
-    //    }
-    //    else
+std::array<StructureOutputBlockMatrix, 3> NuTo::NewmarkDirect::EvaluateHessians()
+{
+    const auto& dofStatus = mStructure->GetDofStatus();
+    std::array<StructureOutputBlockMatrix, 3> hessians = {StructureOutputBlockMatrix(dofStatus, true),
+                                                          StructureOutputBlockMatrix(dofStatus), mHessian2};
+    StructureOutputMap evalHessians;
+    evalHessians[eStructureOutput::HESSIAN0] = &hessians[0];
+    if (mStructure->GetNumTimeDerivatives() >= 1)
     {
-        // since rHessian0 will change in the next iteration, the rHessian0 will be the hessian for the solver
-        if (mStructure->GetNumTimeDerivatives() >= 1)
-            rHessian_dt0.AddScal(rHessian_dt1, mGamma / (mBeta * rTimeStep));
+        if (mUseMuDamping)
+            hessians[1] = mDampingMatrix;
+        else
+        {
+            hessians[1].Resize(dofStatus.GetNumActiveDofsMap(), dofStatus.GetNumDependentDofsMap());
+            evalHessians[eStructureOutput::HESSIAN1] = &hessians[1];
+        }
+    }
+
+    auto inputMap = CreateInputMap();
+    mStructure->Evaluate(inputMap, evalHessians);
+    if (mCheckCoefficientMatrix)
+        mStructure->ElementCheckHessian0(1.e-6, 1.e-8);
+
+    return hessians;
+}
+
+
+std::pair<StructureOutputBlockVector, std::array<StructureOutputBlockMatrix, 3>>
```

Comment:
I'm trying to use type aliases sparingly. If I see `std::array<StructureOutputBlockMatrix, 3>` I know what's going on, assuming I know what a `StructureOutputBlockMatrix` is. With `Hessians` I'll have to go look what this new type I've never heard of represents. But I'll bow to the consensus.